### ISSUE TITLE
Cache conflict when executing pdepend in parallel

### ISSUE THREAD

**Issue by bvis** [pdepend/pdepend#193]
From time to time there's a conflict when PDepends tries to flush the cache and there's another execution in parallel. The output I receive in a Jenkins job is something similar to:

````
phpmd-ci:
PHP Warning:  unlink(/var/lib/jenkins/.pdepend/cr/cr7ujpijhr4-ad7ikjyoew0-00-baila17gjw0.5.6.tokens): No such file or directory in /var/lib/jenkins/jobs/MYPROJECT/workspace/vendor/pdepend/pdepend/src/main/php/PDepend/Util/Cache/Driver/File/FileCacheDirectory.php on line 220
PHP Stack trace:
PHP   1. {main}() /var/lib/jenkins/jobs/MYPROJECT/workspace/vendor/phpmd/phpmd/src/bin/phpmd:0
PHP   2. PHPMD\TextUI\Command::main() /var/lib/jenkins/jobs/MYPROJECT/workspace/vendor/phpmd/phpmd/src/bin/phpmd:122
PHP   3. PHPMD\TextUI\Command->run() /var/lib/jenkins/jobs/MYPROJECT/workspace/vendor/phpmd/phpmd/src/main/php/PHPMD/TextUI/Command.php:175
PHP   4. PHPMD\PHPMD->processFiles() /var/lib/jenkins/jobs/MYPROJECT/workspace/vendor/phpmd/phpmd/src/main/php/PHPMD/TextUI/Command.php:134
PHP   5. PHPMD\Parser->parse() /var/lib/jenkins/jobs/MYPROJECT/workspace/vendor/phpmd/phpmd/src/main/php/PHPMD/PHPMD.php:222
PHP   6. PDepend\Engine->analyze() /var/lib/jenkins/jobs/MYPROJECT/workspace/vendor/phpmd/phpmd/src/main/php/PHPMD/Parser.php:123
PHP   7. PDepend\Engine->performParseProcess() /var/lib/jenkins/jobs/MYPROJECT/workspace/vendor/pdepend/pdepend/src/main/php/PDepend/Engine.php:318
PHP   8. PDepend\Util\Cache\CacheFactory->create() /var/lib/jenkins/jobs/MYPROJECT/workspace/vendor/pdepend/pdepend/src/main/php/PDepend/Engine.php:558
PHP   9. PDepend\Util\Cache\CacheFactory->createCache() /var/lib/jenkins/jobs/MYPROJECT/workspace/vendor/pdepend/pdepend/src/main/php/PDepend/Util/Cache/CacheFactory.php:94
PHP  10. PDepend\Util\Cache\CacheFactory->createFileCache() /var/lib/jenkins/jobs/MYPROJECT/workspace/vendor/pdepend/pdepend/src/main/php/PDepend/Util/Cache/CacheFactory.php:113
PHP  11. PDepend\Util\Cache\Driver\FileCacheDriver->__construct() /var/lib/jenkins/jobs/MYPROJECT/workspace/vendor/pdepend/pdepend/src/main/php/PDepend/Util/Cache/CacheFactory.php:131
PHP  12. PDepend\Util\Cache\Driver\File\FileCacheDirectory->__construct() /var/lib/jenkins/jobs/MYPROJECT/workspace/vendor/pdepend/pdepend/src/main/php/PDepend/Util/Cache/Driver/FileCacheDriver.php:111
PHP  13. PDepend\Util\Cache\Driver\File\FileCacheDirectory->flush() /var/lib/jenkins/jobs/MYPROJECT/workspace/vendor/pdepend/pdepend/src/main/php/PDepend/Util/Cache/Driver/File/FileCacheDirectory.php:79
PHP  14. PDepend\Util\Cache\Driver\File\FileCacheDirectory->flushDirectory() /var/lib/jenkins/jobs/MYPROJECT/workspace/vendor/pdepend/pdepend/src/main/php/PDepend/Util/Cache/Driver/File/FileCacheDirectory.php:189
PHP  15. PDepend\Util\Cache\Driver\File\FileCacheDirectory->flushEntry() /var/lib/jenkins/jobs/MYPROJECT/workspace/vendor/pdepend/pdepend/src/main/php/PDepend/Util/Cache/Driver/File/FileCacheDirectory.php:203
PHP  16. PDepend\Util\Cache\Driver\File\FileCacheDirectory->flushDirectory() /var/lib/jenkins/jobs/MYPROJECT/workspace/vendor/pdepend/pdepend/src/main/php/PDepend/Util/Cache/Driver/File/FileCacheDirectory.php:222
PHP  17. PDepend\Util\Cache\Driver\File\FileCacheDirectory->flushEntry() /var/lib/jenkins/jobs/MYPROJECT/workspace/vendor/pdepend/pdepend/src/main/php/PDepend/Util/Cache/Driver/File/FileCacheDirectory.php:203
PHP  18. unlink() /var/lib/jenkins/jobs/MYPROJECT/workspace/vendor/pdepend/pdepend/src/main/php/PDepend/Util/Cache/Driver/File/FileCacheDirectory.php:220
PHP Warning:  rmdir(/var/lib/jenkins/.pdepend/cr): No such file or directory in /var/lib/jenkins/jobs/MYPROJECT/workspace/vendor/pdepend/pdepend/src/main/php/PDepend/Util/Cache/Driver/File/FileCacheDirectory.php on line 223
PHP Stack trace:
PHP   1. {main}() /var/lib/jenkins/jobs/MYPROJECT/workspace/vendor/phpmd/phpmd/src/bin/phpmd:0
PHP   2. PHPMD\TextUI\Command::main() /var/lib/jenkins/jobs/MYPROJECT/workspace/vendor/phpmd/phpmd/src/bin/phpmd:122
PHP   3. PHPMD\TextUI\Command->run() /var/lib/jenkins/jobs/MYPROJECT/workspace/vendor/phpmd/phpmd/src/main/php/PHPMD/TextUI/Command.php:175
PHP   4. PHPMD\PHPMD->processFiles() /var/lib/jenkins/jobs/MYPROJECT/workspace/vendor/phpmd/phpmd/src/main/php/PHPMD/TextUI/Command.php:134
PHP   5. PHPMD\Parser->parse() /var/lib/jenkins/jobs/MYPROJECT/workspace/vendor/phpmd/phpmd/src/main/php/PHPMD/PHPMD.php:222
PHP   6. PDepend\Engine->analyze() /var/lib/jenkins/jobs/MYPROJECT/workspace/vendor/phpmd/phpmd/src/main/php/PHPMD/Parser.php:123
PHP   7. PDepend\Engine->performParseProcess() /var/lib/jenkins/jobs/MYPROJECT/workspace/vendor/pdepend/pdepend/src/main/php/PDepend/Engine.php:318
PHP   8. PDepend\Util\Cache\CacheFactory->create() /var/lib/jenkins/jobs/MYPROJECT/workspace/vendor/pdepend/pdepend/src/main/php/PDepend/Engine.php:558
PHP   9. PDepend\Util\Cache\CacheFactory->createCache() /var/lib/jenkins/jobs/MYPROJECT/workspace/vendor/pdepend/pdepend/src/main/php/PDepend/Util/Cache/CacheFactory.php:94
PHP  10. PDepend\Util\Cache\CacheFactory->createFileCache() /var/lib/jenkins/jobs/MYPROJECT/workspace/vendor/pdepend/pdepend/src/main/php/PDepend/Util/Cache/CacheFactory.php:113
PHP  11. PDepend\Util\Cache\Driver\FileCacheDriver->__construct() /var/lib/jenkins/jobs/MYPROJECT/workspace/vendor/pdepend/pdepend/src/main/php/PDepend/Util/Cache/CacheFactory.php:131
PHP  12. PDepend\Util\Cache\Driver\File\FileCacheDirectory->__construct() /var/lib/jenkins/jobs/MYPROJECT/workspace/vendor/pdepend/pdepend/src/main/php/PDepend/Util/Cache/Driver/FileCacheDriver.php:111
PHP  13. PDepend\Util\Cache\Driver\File\FileCacheDirectory->flush() /var/lib/jenkins/jobs/MYPROJECT/workspace/vendor/pdepend/pdepend/src/main/php/PDepend/Util/Cache/Driver/File/FileCacheDirectory.php:79
PHP  14. PDepend\Util\Cache\Driver\File\FileCacheDirectory->flushDirectory() /var/lib/jenkins/jobs/MYPROJECT/workspace/vendor/pdepend/pdepend/src/main/php/PDepend/Util/Cache/Driver/File/FileCacheDirectory.php:189
PHP  15. PDepend\Util\Cache\Driver\File\FileCacheDirectory->flushEntry() /var/lib/jenkins/jobs/MYPROJECT/workspace/vendor/pdepend/pdepend/src/main/php/PDepend/Util/Cache/Driver/File/FileCacheDirectory.php:203
PHP  16. rmdir() /var/lib/jenkins/jobs/MYPROJECT/workspace/vendor/pdepend/pdepend/src/main/php/PDepend/Util/Cache/Driver/File/FileCacheDirectory.php:223
Directory name must not be empty.
Result: 1
````
One possible solution would be to isolate the cache of PDepend not just per user as it is now, but for process. Do this is so easy, you just need to change in _PDepend\DependencyInjection\Configuration_(88):
````
->scalarNode('location')->defaultValue($home . '/.pdepend')->end()
````
By something like:
````
->scalarNode('location')->defaultValue($home . '/.pdepend/'.getmypid())->end()
````
But this has an undesired side effect, the implemented cache flush will not work correctly because each cache is saved in a separated directory that will probably never cleaned up by any later PDepend execution. It should be changed to check not just this cache directory that includes the PID, but the generic cache directory.

**Comment by bvis** [comment]
I think this is related with: #163
I'll reactivate my job execution parallelization to check if it fixed the problem.

**Comment by manuelpichler** [comment]
Very sad, but this is an issue that came up several times before :( Using the pid is not a good idea because then pdepend would floud your file system.

Since the cache cleanup is not a critical thing, I will now try to fix this issue by letting ```unlink``` and ```rmdir``` silently fail.

**Comment by bvis** [comment]
Hi Manuel,

I can tell you that after the #163 fix I have not seen this problem again. I can't ensure that it can't be produced again but it's good to know!

**Comment by manuelpichler** [comment]
FIxed in a4e20ff
